### ISSUE TITLE
Reflection fails for enums with 2 or more associated values

### ISSUE THREAD

**Issue by michael404** [pointfreeco/swift-case-paths#49]
Reflection seems to fail for enums with 2 or more associated values.

Can be reproduced by opening a new Xcode project (Xcode 12.5), adding swift-case-paths as a dependency (0.6.2) and running the following code in main:

```swift
import CasePaths
enum E { case c(Int, Int) }
/E.c
```
This triggers a "value must be a valid enum case" fatalError from the new reflection code. Removing one of the ints resolves the issue.

I would expect this to work and produce a CasePath<E, (Int, Int)>. Is this not meant to be supported?

**Comment by stephencelis** [comment]
@michael404 This fatal error is new, but the behavior is unfortunately long-standing and is related to Swift overload resolution.

TL;DR: you can use an explicit type annotation, or use the case path where the types can be inferred, to fix the problem:

```swift
let _: CasePath<E, (Int, Int)> = /E.c
``

What we've found in practice is that case paths are almost always used in situations where there's no ambiguity in the types involved. Can you give us more information as to your use case?

---

As for explaining, and exploring fixes...

There are 2 main "kinds" of case paths: those for enum cases with associated values, and those for cases without. For example, this enum:

```swift
enum A {
  case b(Bool)
  case c
}
```

Can create case paths for `b` and `c`:

```swift
_: CasePath<A, Bool> = /A.b
_: CasePath<A, Void> = /A.c
```

This is made possible with some overloads, but unfortunately it means that if we have an enum with 2+ associated values, the default promotion goes through the `Void` overload, where it thinks the "root" is the embed function:

```swift
enum B {
  case c(Int, Int)
}

let path = /B.c
// path is type 'CasePath<(Int, Int) -> B, Void>'
```

It is possible to convince the compiler that `/(A, B) -> C` should be `CasePath<C, (A, B)>` and so on, but it requires _even more overloads_, in fact an overload for each number of associated values, which isn't ideal. There may also be a way of using `@_disfavoredOverload` to work around the problem, but we haven't figured out a solution with that yet. We're also open to any other solutions that folks come up with!

**Comment by michael404** [comment]
@stephencelis: Thanks for the detailed info - that makes total sense. To be honest I don't have a particular use case - I was just playing around with CasePaths in a playground when I discovered this.

It is a little bit sad that there is no way to restrict `Root` to be an Enum. Not sure if the recently accepted magic `@_marker Actor` protocol could be be president for a pitch to add a simular magic `Enum` protocol. If I understand correctly, this would all be solved if we could restrict `Root` to be an enum.

In the meantime, I guess in theory we could use reflection to discover that `Root` is a function and `Value` is `Void` at runtime and give a more helpful assertionFailure for this particular situation. My issue in debugging this issue was that the tried and true `print(type(of: /E.c))` also failed and that the "value must be a valid enum case" message did not make sense to me.

**Comment by stephencelis** [comment]
Going to close this for now since it's unlikely to occur in a production situation, and because the only solution we know of requires a lot of overloads. If anyone knows a fix that does _not_ require overloads, though, we'd love a PR!
